internal/controller/helm.go

```diff
@@ -125,6 +144,75 @@ func (r *UpgradePlanReconciler) createHelmChart(ctx context.Context, upgradePlan
 	return r.createObject(ctx, upgradePlan, chart)
 }
 
+func mergeHelmValues(installedValues any, releaseValues, userValues *apiextensionsv1.JSON) ([]byte, error) {
+	values := map[string]any{}
+
+	switch installed := installedValues.(type) {
+	case string:
+		if installed != "" {
+			if err := json.Unmarshal([]byte(installed), &values); err != nil {
+				return nil, fmt.Errorf("unmarshaling installed chart values: %w", err)
+			}
+		}
+	case map[string]interface{}:
+		if len(installed) != 0 {
+			maps.Copy(values, installed)
+		}
+	default:
+		return nil, fmt.Errorf("unexpected type %T of installed values", installedValues)
+	}
+
+	if releaseValues != nil && len(releaseValues.Raw) > 0 {
+		var v map[string]any
+
+		if err := json.Unmarshal(releaseValues.Raw, &v); err != nil {
+			return nil, fmt.Errorf("unmarshaling additional release values: %w", err)
+		}
+
+		values = mergeMaps(values, v)
+	}
+
+	if userValues != nil && len(userValues.Raw) > 0 {
+		var v map[string]any
+
+		if err := json.Unmarshal(userValues.Raw, &v); err != nil {
+			return nil, fmt.Errorf("unmarshaling additional user values: %w", err)
+		}
+
+		values = mergeMaps(values, v)
+	}
+
+	if len(values) == 0 {
+		return nil, nil
+	}
+
+	v, err := json.Marshal(values)
+	if err != nil {
+		return nil, fmt.Errorf("marshaling chart values: %w", err)
+	}
+
+	return v, nil
+}
+
+func mergeMaps(m1, m2 map[string]any) map[string]any {
+	out := make(map[string]any, len(m1))
+	for k, v := range m1 {
+		out[k] = v
+	}
+
+	for k, v := range m2 {
+		if v, ok := v.(map[string]any); ok {
```

Comment:
Nit: naming is a bit hard to follow, can we have something like:
```suggestion
		if inner, ok := v.(map[string]any); ok {
```